src/blob.rs

```diff
@@ -1,86 +1,342 @@
-use std::path::Path;
+use futures::{
+    future::Future,
+    task::{Poll, Waker},
+};
+use std::{
+    collections::BTreeMap,
+    fs, io,
+    os::unix::fs::FileExt,
+    path::{Path, PathBuf},
+    pin::Pin,
+};
 
-use crate::record::Record;
+use crate::record::{self, Record};
+
+const BLOB_MAGIC_BYTE: u64 = 0xdeaf_abcd;
+
+type Result<T> = std::result::Result<T, Error>;
 
 /// A `Blob` struct for performing of database,
-#[derive(Debug, Default)]
-pub struct Blob<T> {
+#[derive(Debug)]
+pub struct Blob {
     header: Header,
-    records: Vec<T>, // @TODO needs verification, created to yield generic T up
+    index: Index,
+    name: FileName,
+    file: fs::File,
+    current_offset: u64,
 }
 
-/// # Description
-/// # Examples
 #[derive(Debug, Default)]
-struct Header {
-    magic_byte: u64,
-    version: u32,
-    key_size: u32,
-    flags: u64,
+struct Index {
+    headers: BTreeMap<Vec<u8>, record::Header>,
 }
```

Comment:
I suggest using the range index additionally. It would allow to reject not presented entries faster.

Reply:
@ikopylov add issue for future plans